lib/pushmi_pullyu/aip.rb

```diff
@@ -12,13 +12,16 @@ def create(entity)
 
     aip_directory = "#{PushmiPullyu.options[:workdir]}/#{entity[:uuid]}"
     aip_filename = "#{aip_directory}.tar"
+    begin
+      PushmiPullyu::AIP::Downloader.new(entity, aip_directory).run
+      PushmiPullyu::AIP::Creator.new(entity[:uuid], aip_directory, aip_filename).run
 
-    PushmiPullyu::AIP::Downloader.new(entity, aip_directory).run
-    PushmiPullyu::AIP::Creator.new(entity[:uuid], aip_directory, aip_filename).run
-
-    yield aip_filename, aip_directory
-
-    FileUtils.rm_rf(aip_filename) if File.exist?(aip_filename)
-    FileUtils.rm_rf(aip_directory) if File.exist?(aip_directory)
+      yield aip_filename, aip_directory
+    # Leave expection handling to logic at a higher level
```

Comment:
I'm not sure what this comment is referring to?  Exceptions will be handled by the ensure block.  Which logic at what level?

Reply:
I am using a ``begin`` block to add an ``ensure`` block in order to remove the files if something goes wrong. I am purposely not adding a ``rescue`` block that only re-raises the exception to be handled elsewhere. 

In our current version this is watched here:

https://github.com/ualbertalib/pushmi_pullyu/blob/c24aba37da054fe542c9293b96e5aa9930b557fb/lib/pushmi_pullyu/cli.rb#L203-L209

That piece of code is where the next work will be added to re-add the ids of the items that were not processed correctly.

I should make that comment more descriptive.

Reply:
To continue with our conversation:

The ensure part of the creation method is a sort of garbage collection for temporary files, there is no rescue block there, so essentially i am just re-raising the exception.

if an exception occurs it goes up to the ``run_preservation_cycle`` method in ``cli``

https://github.com/ualbertalib/pushmi_pullyu/blob/25f2ed564ab3ed1d75cca3ec7532aec36820c904/lib/pushmi_pullyu/cli.rb#L184-L210

The following piece of code is where the exception is caught

https://github.com/ualbertalib/pushmi_pullyu/blob/25f2ed564ab3ed1d75cca3ec7532aec36820c904/lib/pushmi_pullyu/cli.rb#L193-L209

it is in that rescue ``Exception => e`` block where I will re-add to redis the id of the item that failed to be preserved

It may be clearer to re-raise the exception. The only benefit of not explicitly adding that part is to have fewer lines. If fewer lines come at the cost of legibility, I will add that rescue block. @pgwillia, do you think this would be a better approach?

An example of how the exception is handled can be more apparent in its test

https://github.com/ualbertalib/pushmi_pullyu/blob/25f2ed564ab3ed1d75cca3ec7532aec36820c904/spec/pushmi_pullyu/aip_spec.rb#L58-L85

Specifically on this piece of code:

https://github.com/ualbertalib/pushmi_pullyu/blob/25f2ed564ab3ed1d75cca3ec7532aec36820c904/spec/pushmi_pullyu/aip_spec.rb#L71-L84

Here I pass a block that simply raises a ``StandardError`` and we check the files were removed by the ``ensure`` block.

Happy to hear your thoughts on this.